File: 2807.insert-greatest-common-divisors-in-linked-list.py
#
# @lc app=leetcode id=2807 lang=python
#
# [2807] Insert Greatest Common Divisors in Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def insertGreatestCommonDivisors(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        node = head
        def gcd(a, b):
            return a if b == 0 else gcd(b, a % b)
        while node.next:
           node.next, node = ListNode(gcd(node.val, node.next.val), node.next), node.next
        return head
        
# @lc code=end

File: 2807.insert-greatest-common-divisors-in-linked-list.java
/*
 * @lc app=leetcode id=2807 lang=java
 *
 * [2807] Insert Greatest Common Divisors in Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode insertGreatestCommonDivisors(ListNode head) {
        if (head.next == null) {
      return head;
    }
    ListNode prev = head;
    ListNode curr = head.next;
    while (curr != null) {
      int a = prev.val;
      int b = curr.val;
      int min = Math.min(a,b);
      int gcd = 1;
      if (a % min == 0 && b % min == 0) {
        gcd = min;
      }
      else {
        for (int i = 2; i < min/2 + 1; i++) {
          if (a % i == 0 && b % i == 0) {
            gcd = i;
          }
        }
      }
      ListNode node = new ListNode(gcd);
      node.next = curr;
      prev.next = node;
      prev = curr;
      curr = curr.next;
    }
    return head;
    }
}
// @lc code=end

